### PR TITLE
bugfix: spsa works with latest numpy

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
SPSA tests have been failing with numpy >= 1.25, and we couldn't find out why. The true reason is a difference in behaviour with ~autograd.numpy's implementation of divide~ `pennylane.numpy.tensor.__array_ufunc__`, but I can explain that below.

**Description of the Change:**
- ~Change the custom sampler in SPSA tests to use vanilla numpy~
- Cast the `where` argument passed to `qml.math.divide` to a vanilla python list to avoid side effects
- Unpin numpy in docker builds now that this is fixed

**Benefits:**
Our tests pass with the latest numpy!

**Possible Drawbacks:**
N/A

[sc-47695]

### Investigation Part 1

We use `qml.math.divide` on the results of a sampler call [in spsa_gradient.py](https://github.com/PennyLaneAI/pennylane/blob/4e1f7495853df620852a50ba2daa9bceeb4bd86c/pennylane/gradients/spsa_gradient.py#L332-L335). The custom sampler in these tests was using pennylane.numpy instead of vanilla numpy, but they appear to behave differently:

```pycon
>>> div = lambda x: qml.math.divide(1, x, where=(x != 0), out=qml.math.zeros_like(x))
>>> good = np.array([0., 1.])
>>> bad = pnp.array([0., 1.])
>>> div(good), div(bad)
(array([0., 1.]), tensor([0., 0.], requires_grad=True))
```

What's happening here? For each term of the denominator (`x`), if the `where` clause evaluates True, `divide` evaluates the quotient for that term. If it instead evaluated False (in other words, we're about to divide by zero), it gets the value from `out` as the effective default. For some reason, `out` must match the input dimension, so I can't change this to be `qml.math.convert_like(0., x)`.

With vanilla numpy, it only replaces values with the `out`-defaults when the denominator was zero, but with pennylane numpy, it just returns `out`. I'm just avoiding using pennylane-numpy in tests here because it's unnecessary. naughty autograd.

### Investigation Part 2

See my comment below. Basically, passing `where` to numpy functions now behaves differently if the value implements `__array_ufunc__`, and pennylane-numpy tensors do in fact override it.

### CODE REVIEW

**Review by lillian542 — APPROVED** [review]
What a weird problem! Glad this fixes it here, but are we generally unconcerned about there being an unexpected discrepancy between pennylane numpy and vanilla numpy for new versions of numpy? I don't feel entirely unconcerned.

Regardless, this looks good to me! 🎉

**Comment by timmysilv** [comment]
I'm a little concerned too, but idk if I want to open that can of worms only for this particular case

**Comment by timmysilv** [comment]
@lillian542 re-requested review because I decided to just fix the source code in case someone else provides a sampler that returns a funky value (aka a value that implements/overrides `__array_ufunc__`)

**Review by mudit2812 — APPROVED** [review]

**Review by lillian542 — APPROVED** [review]
Seconding that a comment in the file would be nice to have, the purpose of casting to a list is very unclear without context. Looks good to me!

File: pennylane/gradients/spsa_gradient.py [hunk 1]
```diff
@@ -331,7 +331,7 @@ def spsa_grad(
     for idx_rep in range(num_directions):
         direction = sampler(indices, num_trainable_params, idx_rep, rng=sampler_rng)
         inv_direction = qml.math.divide(
-            1, direction, where=(direction != 0), out=qml.math.zeros_like(direction)
+            1, direction, where=list(direction != 0), out=qml.math.zeros_like(direction)
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/gradients/spsa_gradient.py:336]: done!

**Review by albi3ro — APPROVED** [review]
